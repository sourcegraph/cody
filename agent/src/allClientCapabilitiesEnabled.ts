import type { ClientCapabilities } from '@sourcegraph/cody-shared'

export const allClientCapabilitiesEnabled: ClientCapabilities = {
    progressBars: 'enabled',
    edit: 'enabled',
    editWorkspace: 'enabled',
    showDocument: 'enabled',
    codeLenses: 'enabled',
    showWindowMessage: 'request',
    ignore: 'enabled',
    codeActions: 'enabled',
    secrets: 'client-managed',
    authentication: 'enabled',
}
