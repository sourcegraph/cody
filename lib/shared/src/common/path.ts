import type { URI } from 'vscode-uri'

import { isWindows as _isWindows, isMacOS } from './platform'

interface PathFunctions {
    /**
     * All but the last element of path, or "." if that would be the empty path.
     */
    dirname: (path: string) => string

    /**
     * The last element of path, or "" if path is empty.
     * @param path the path to operate on
     * @param suffix optional suffix to remove
     */
    basename: (path: string, suffix?: string) => string

    /** The extension of path, including the last '.'. */
    extname: (path: string) => string

    /**
     * The relative path from {@link from} to {@link to}.
     */
    relative: (from: string, to: string) => string

    resolve: (...paths: string[]) => string

    /** Path separator. */
    separator: PathSeparator
}

/** For file system paths on Windows ('\' separators, drive letters, case-insensitive). */
export const windowsFilePaths: PathFunctions = pathFunctions(true, '\\', false)

/** For POSIX file system paths ('/' separators, case-sensitive). */
export const posixFilePaths: PathFunctions = pathFunctions(false, '/', true)

/**
 * Get the {@link PathFunctions} to use for the given URI's path ('/' separators, drive letters/case-sensitivity depend on `isWindows`).
 */
export function pathFunctionsForURI(uri: URI, isWindows = _isWindows()): PathFunctions {
    // URIs are always forward slashes even on Windows.
    const sep = '/'
    return uri.scheme === 'file' && isWindows
        ? // Like windowsFilePaths but with forward slashes.
          pathFunctions(true, sep, false)
        : posixFilePaths
}

export function defaultPathFunctions(): PathFunctions {
    const isWindows = _isWindows()
    return pathFunctions(isWindows, isWindows ? '\\' : '/', isMacOS())
}

// I don't like reimplementing this here, but it's the best option because: (1) using Node's `path`
// module requires us to configure the bundler right or risk a subtle bug (and path-browserify
// doesn't have `path/win32` support, so we'd need multiple underlying packages); and (2)
// `vscode-uri` is hard to test because it has a global constant for the current platform set at
// init time.
//
// note: case-insensitive for Windows is not strictly sound, it's possible to have case-sensitive
// volumes on Windows (and case-insensitive on other platforms), however it's impossible to tell from
// a URI/path alone whether that's the case and VS Code already assumes based on platform, so things
// are unfortunately already broken if you don't use the "defaults".
// - https://github.com/microsoft/vscode/issues/94307
// - https://github.com/microsoft/vscode/issues/177925
function pathFunctions(isWindows: boolean, sep: '\\' | '/', caseSensitive: boolean): PathFunctions {
    const f: PathFunctions = {
        dirname(path: string): string {
            if (path === '') {
                return '.'
            }
            if (isWindows && isDriveLetter(path)) {
                return path
            }
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            if (path === '') {
                return sep
            }
            const i = path.lastIndexOf(sep)
            if (i === -1) {
                return '.'
            }
            if (i === 0) {
                return sep
            }
            path = path.slice(0, i)
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            return path
        },
        basename(path: string, suffix?: string): string {
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return ''
            }
            path = path.split(sep).at(-1) ?? ''
            if (suffix && path.endsWith(suffix)) {
                path = path.slice(0, -suffix.length)
            }
            return path
        },
        extname(path: string): string {
            const basename = f.basename(path)
            const i = basename.lastIndexOf('.')
            if (i === 0 || i === -1) {
                return ''
            }
            return basename.slice(i)
        },
        relative(from, to) {
            // For relative paths, output separator is always based on platform, even
            // if the input (sep) is forward slash/URI.
            return relative(from, to, sep, isWindows ? '\\' : '/', caseSensitive)
        },
        resolve(...paths) {
            return paths.join(sep)
        },
        separator: sep,
    }
    return f
}

function isDriveLetter(path: string): boolean {
    return /^[A-Za-z]:$/.test(path)
}

/** The relative path from {@link from} to {@link to}. */
function relative(
    from: string,
    to: string,
    inputSeparator: PathSeparator,
    outputSeparator: PathSeparator,
    caseSensitive: boolean
): string {
    function equalSegments(a: string, b: string) {
        return caseSensitive ? a === b : a.toLowerCase() === b.toLowerCase()
    }

    // Normalize extra slashes.
    if (inputSeparator === '/') {
        from = from.replaceAll(/\/{2,}/g, '/')
        to = to.replaceAll(/\/{2,}/g, '/')
    } else {
        from = from.replaceAll(/\\{2,}/g, '\\')
        to = to.replaceAll(/\\{2,}/g, '\\')
    }

    // Trim trailing slashes.
    if (from !== inputSeparator && from.endsWith(inputSeparator)) {
        from = from.slice(0, -1)
    }
    if (to !== inputSeparator && to.endsWith(inputSeparator)) {
        to = to.slice(0, -1)
    }

    if (equalSegments(from, to)) {
        return ''
    }

    // From a relative to an absolute path, the absolute path dominates.
    if (!from.startsWith(inputSeparator) && to.startsWith(inputSeparator)) {
        return to
    }

    const fromParts = from === inputSeparator ? [''] : from.split(inputSeparator)
    const toParts = to === inputSeparator ? [''] : to.split(inputSeparator)

    // Find the common root.
    let commonLength = 0
    while (
        commonLength < fromParts.length &&
        commonLength < toParts.length &&
        equalSegments(fromParts[commonLength], toParts[commonLength])
    ) {
        commonLength++
    }

    // Calculate the number of directory traversals needed.
    const traversals = fromParts.length - commonLength

    // Build the relative path.
    const relativePath: string[] = []
    for (let i = 0; i < traversals; i++) {
        relativePath.push('..')
    }

    // Add the non-common path parts from the 'to' path.
    relativePath.push(...toParts.slice(commonLength))

    return relativePath.join(outputSeparator)
}

type PathSeparator = '\\' | '/'
