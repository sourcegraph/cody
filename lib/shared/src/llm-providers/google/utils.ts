import type { Content, InlineDataPart, Part } from '@google/generative-ai'
import type { Message } from '../..'
import { getMessageImageUrl } from '../completions-converter'

/**
 * Constructs the messages array for the Gemini API, including handling InlineDataPart for media.
 */
export async function constructGeminiChatMessages(messages: Message[]): Promise<Content[]> {
    const contents: Content[] = []

    // Map speaker types to Gemini API roles
    const roleMap: Record<string, 'user' | 'model' | 'system'> = {
        human: 'user',
        assistant: 'model',
        system: 'system',
    }

    for (const message of messages) {
        const role = roleMap[message.speaker] || 'user'
        const parts: Part[] = []

        // Skip if this would create consecutive messages from the same role
        const lastContent = contents[contents.length - 1]
        if (lastContent?.role === role) {
            continue
        }

        // Process message content parts
        if (message.content?.length) {
            for (const part of message.content) {
                if (part.type === 'text' && part.text?.length) {
                    parts.push({ text: part.text })
                }
                const { data, mimeType } = getMessageImageUrl(part)
                if (data && mimeType) {
                    parts.push({
                        inlineData: { mimeType, data: data.replace(/data:[^;]+;base64,/, '') },
                    } satisfies InlineDataPart)
                }
            }
        }

        // Add message text if present
        if (message.text?.length) {
            parts.push({ text: message.text.toString() })
        }

        // Add content if there are parts
        if (parts.length > 0) {
            contents.push({ role, parts })
        }
    }

    // Remove trailing model message if present
    if (contents.length > 0 && contents[contents.length - 1].role === 'model') {
        contents.pop()
    }

    return contents
}
