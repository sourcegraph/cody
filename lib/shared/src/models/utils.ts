import type { ClientConfiguration } from '../configuration'
import { logDebug } from '../logger'
import { ANSWER_TOKENS } from '../prompt/constants'
import type { CodyLLMSiteConfiguration } from '../sourcegraph-api/graphql/client'
import {
    CHAT_INPUT_TOKEN_BUDGET,
    CHAT_OUTPUT_TOKEN_BUDGET,
    EXTENDED_CHAT_INPUT_TOKEN_BUDGET,
    EXTENDED_USER_CONTEXT_TOKEN_BUDGET,
} from '../token/constants'
import type { Model } from './model'
import type { ModelRef, ModelRefStr } from './modelsService'
import { ModelTag } from './tags'
import type { ModelContextWindow } from './types'

export function getProviderName(name: string): string {
    const providerName = name.toLowerCase()
    switch (providerName) {
        case 'anthropic':
            return 'Anthropic'
        case 'openai':
            return 'OpenAI'
        case 'ollama':
            return 'Ollama'
        case 'google':
            return 'Google'
        default:
            return providerName
    }
}

/**
 * Gets the provider and title from a model ID string.
 */
export function getModelInfo(modelID: string): {
    provider: string
    title: string
} {
    const [providerID, ...rest] = modelID.split('/')
    const provider = getProviderName(providerID)
    const title = (rest.at(-1) || '').replace(/-/g, ' ')
    return { provider, title }
}

export function isCodyProModel(model: Model): boolean {
    return modelHasTag(model, ModelTag.Pro)
}

export function isWaitlistModel(model: Model): boolean {
    return modelHasTag(model, ModelTag.Waitlist) || modelHasTag(model, ModelTag.OnWaitlist)
}

export function isCustomModel(model: Model): boolean {
    return (
        modelHasTag(model, ModelTag.Local) ||
        modelHasTag(model, ModelTag.Dev) ||
        modelHasTag(model, ModelTag.BYOK)
    )
}

function modelHasTag(model: Model, modelTag: ModelTag): boolean {
    return model.tags.includes(modelTag)
}

export function toModelRefStr(modelRef: ModelRef): ModelRefStr {
    const { providerId, apiVersionId, modelId } = modelRef
    return `${providerId}::${apiVersionId}::${modelId}`
}

/**
 * Get the context window for the given chat model and configuration overwrites in an enterprise environment.
 *
 * @param chatModel - The chat model to get the context window for.
 * @param configOverwrites - The configuration overwrites to apply.
 * @returns The context window for the given chat model and configuration overwrites.
 */
export function getEnterpriseContextWindow(
    chatModel: string,
    configOverwrites: CodyLLMSiteConfiguration,
    configuration: Pick<ClientConfiguration, 'providerLimitPrompt'>
): ModelContextWindow {
    const { chatModelMaxTokens, smartContextWindow } = configOverwrites
    // Starts with the default context window.
    let contextWindow: ModelContextWindow = {
        input: chatModelMaxTokens ?? CHAT_INPUT_TOKEN_BUDGET,
        output: getEnterpriseOutputLimit(chatModel),
    }

    // Use extended context window for models that support smart context when enabled.
    if (smartContextWindow && isModelWithExtendedContextWindowSupport(chatModel)) {
        contextWindow = {
            input: EXTENDED_CHAT_INPUT_TOKEN_BUDGET,
            output: CHAT_OUTPUT_TOKEN_BUDGET,
            context: { user: EXTENDED_USER_CONTEXT_TOKEN_BUDGET },
        }
    }

    return applyLocalTokenLimitOverwrite(configuration, chatModel, contextWindow)
}

/**
 * Applies a local token limit overwrite to the given context window if configured.
 * If the configured limit is lower than the default, it will be applied to the input of the context window.
 * If the configured limit is invalid, an error will be logged.
 *
 * @param chatModel The chat model for which the token limit is being applied.
 * @param contextWindow The context window to apply the token limit overwrite to.
 * @returns The updated context window with the token limit overwrite applied, or the original context window if no valid overwrite is configured.
 */
function applyLocalTokenLimitOverwrite(
    { providerLimitPrompt }: Pick<ClientConfiguration, 'providerLimitPrompt'>,
    chatModel: string,
    contextWindow: ModelContextWindow
): ModelContextWindow {
    if (providerLimitPrompt && providerLimitPrompt <= contextWindow.input) {
        return { ...contextWindow, input: providerLimitPrompt }
    }

    if (providerLimitPrompt) {
        logDebug(
            'getEnterpriseContextWindow',
            `Invalid token limit configured for ${chatModel}`,
            providerLimitPrompt
        )
    }

    return contextWindow
}

/**
 * Returns true if the given chat model supports extended context windows.
 *
 * @param chatModel - The name of the chat model.
 * @returns True if the chat model supports extended context windows, false otherwise.
 */
const modelWithExpandedWindowSubStrings = [
    'opus',
    'sonnet', // Covers claude 3 sonnet & 3.5 sonnet
    'gemini-1.5',
    'gpt-4o',
    'gpt-4-turbo',
]
function isModelWithExtendedContextWindowSupport(chatModel: string): boolean {
    return modelWithExpandedWindowSubStrings.some(keyword => chatModel.toLowerCase().includes(keyword))
}

// TODO: Currently all enterprise models have a max output limit of
// 1000. We need to support configuring the maximum output limit at an
// instance level. This will allow us to increase this limit whilst
// still supporting models with a lower output limit.
//
// To avoid Enterprise instances being stuck with low token counts, we
// will detect our recommended Cody Gateway models to use a higher limit.
//
// See: https://github.com/sourcegrcaph/cody/issues/3648#issuecomment-2056954101
// See: https://github.com/sourcegraph/cody/pull/4203
function getEnterpriseOutputLimit(model?: string) {
    if (model && isModelWithExtendedContextWindowSupport(model)) {
        return CHAT_OUTPUT_TOKEN_BUDGET
    }
    return ANSWER_TOKENS
}
