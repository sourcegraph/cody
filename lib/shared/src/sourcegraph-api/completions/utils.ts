import { type SerializedChatMessage, contextFiltersProvider } from '../..'
import { serverSupportsPromptCaching } from '../clientConfig'
import type { CompletionParameters, Message, SerializedCompletionParameters } from './types'

/**
 * Serializes the completion parameters by converting the message text to a filtered string using the provided context filters.
 *
 * @param params - The completion parameters to serialize.
 * @returns A serialized version of the completion parameters with the message text filtered.
 */
export async function getSerializedParams(
    params: CompletionParameters
): Promise<SerializedCompletionParameters> {
    return {
        ...params,
        messages: await serializePrompts(params.messages, params.model),
    }
}

async function serializePrompts(
    messages: Message[],
    modelID?: string
): Promise<SerializedChatMessage[]> {
    // NOTE: Some models do not support empty assistant message at the end.
    if (modelID?.startsWith('google/') && messages.at(-1)?.speaker === 'assistant') {
        messages.pop()
    }

    return Promise.all(
        messages.map(async m => {
            const text = await m.text?.toFilteredString(contextFiltersProvider)
            if (serverSupportsPromptCaching() && m.cacheEnabled) {
                return {
                    speaker: m.speaker,
                    content: [
                        {
                            ...m,
                            type: 'text',
                            text: text ?? '',
                            cache_control: { type: 'ephemeral' },
                        },
                    ],
                }
            }
            return {
                ...m,
                text: text,
            }
        })
    )
}
