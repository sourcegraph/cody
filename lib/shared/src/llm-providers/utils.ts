import type { CompletionsModelConfig } from '.'
import { modelsService } from '../models/modelsService'

export function getCompletionsModelConfig(modelID: string): CompletionsModelConfig | undefined {
    const provider = modelsService.getModelByID(modelID)
    if (!provider) {
        return undefined
    }

    const { id, clientSideConfig = {} } = provider
    const { apiKey = '', apiEndpoint, options = {} } = clientSideConfig

    const prefix = provider.provider.toLowerCase() + '/'
    const model = id.startsWith(prefix) ? id.substring(prefix.length) : id
    const stream = Boolean(options?.stream ?? true)

    const { stream: _, ...restOptions } = options || {}

    return {
        model,
        key: apiKey,
        endpoint: apiEndpoint,
        stream,
        options: restOptions,
    }
}
