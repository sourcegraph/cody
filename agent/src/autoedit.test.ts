import path from 'node:path'
import { toMatchImageSnapshot } from 'jest-image-snapshot'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import type * as vscode from 'vscode'
import { sleep } from '../../vscode/src/completions/utils'
import { TESTING_CREDENTIALS } from '../../vscode/src/testutils/testing-credentials'
import { TestClient } from './TestClient'
import { TestWorkspace } from './TestWorkspace'
import { allClientCapabilitiesEnabled } from './allClientCapabilitiesEnabled'
import type {
    AutocompleteEditItem,
    AutocompleteResult,
    ExtensionConfiguration,
    Position,
} from './protocol-alias'

expect.extend({ toMatchImageSnapshot })

describe('Autoedit', () => {
    const workspace = new TestWorkspace(path.join(__dirname, '__tests__', 'autoedit'))
    const clientConfiguration: Partial<ExtensionConfiguration> = {
        suggestionsMode: 'auto-edit (Experimental)',
    }

    beforeAll(async () => {
        await workspace.beforeAll()
    })

    afterAll(async () => {
        await workspace.afterAll()
    })

    describe('autoedit - completions', () => {
        const client = TestClient.create({
            workspaceRootUri: workspace.rootUri,
            name: 'autoedit-completions',
            credentials: TESTING_CREDENTIALS.enterprise,
            extraConfiguration: {
                experimentalAutoEditEnabled: true,
            },
            capabilities: {
                ...allClientCapabilitiesEnabled,
                autoedit: 'enabled',
                autoeditInlineDiff: 'insertions-and-deletions',
                autoeditAsideDiff: 'none',
            },
        })

        beforeAll(async () => {
            await client.beforeAll(clientConfiguration)
        })

        afterAll(async () => {
            await client.afterAll()
        })

        it('autocomplete/execute (non-empty result)', async () => {
            const uri = workspace.file('src', 'sum.ts')
            await client.openFile(uri)

            // Set a small visibility delay for testing purposes.
            const visibilityDelay = 100
            await client.request('testing/autocomplete/setCompletionVisibilityDelay', {
                delay: visibilityDelay,
            })

            // Generate completions for the current cursor position.
            const result = await client.request('autocomplete/execute', {
                uri: uri.toString(),
                position: { line: 1, character: 4 },
                triggerKind: 'Automatic',
            })
            const completionID = result.inlineCompletionItems[0].id

            // Tell completion provider that the completion was shown to the user.
            client.notify('autocomplete/completionSuggested', { completionID })

            // Wait for only half of the time required to mark the completion as read.
            await sleep(visibilityDelay / 2)

            // Change the cursor position discarding the current completion.
            // Since the completion was visible only `visibilityDelay / 2` it should not be marked as read.
            await client.changeFile(uri, {
                selection: {
                    start: { line: 0, character: 0 },
                    end: { line: 0, character: 0 },
                },
            })

            // Now the completion visibility timeout is elapsed but because we discarded it, it should not be
            // marked as read because we moved the cursor on the previous step and discarded the completion.
            await client.request('testing/autocomplete/awaitPendingVisibilityTimeout', null)

            const autoeditEvent = await client.request('testing/autocomplete/autoeditEvent', {
                completionID,
            })
            expect(autoeditEvent?.phase).toBe('rejected')
            expect(autoeditEvent?.read).toBe(false)

            // The LLM provided with a completion result.
            expect(result.inlineCompletionItems.length).toBeGreaterThan(0)
            expect(result.decoratedEditItems.length).toBe(0)

            const texts = result.inlineCompletionItems.map(item => item.insertText)
            expect(texts).toMatchInlineSnapshot(
                `
              [
                "    return a + b;",
              ]
            `
            )
        }, 10_000)

        it('autocomplete/execute multiline (non-empty result)', async () => {
            // Open merge sort with a full algorithm implementation to add it to the context.
            // Otherwise there's a higher chance that model won't complete the code because of all
            // the functions in the workspace missing implementations.
            const uri1 = workspace.file('src', 'mergeSort.ts')
            await client.openFile(uri1)

            const uri = workspace.file('src', 'bubbleSort.ts')
            await client.openFile(uri)

            const result = await client.request('autocomplete/execute', {
                uri: uri.toString(),
                position: { line: 1, character: 4 },
                triggerKind: 'Automatic',
            })
            const completionID = result.inlineCompletionItems[0].id

            // Tell completion provider that the completion was shown to the user.
            client.notify('autocomplete/completionSuggested', { completionID })
            // Wait for the completion visibility timeout we use to ensure users read a completion.
            await client.request('testing/autocomplete/awaitPendingVisibilityTimeout', null)
            client.notify('autocomplete/completionAccepted', { completionID })

            const autoeditEvent = await client.request('testing/autocomplete/autoeditEvent', {
                completionID,
            })
            expect(autoeditEvent?.phase).toBe('accepted')
            expect(autoeditEvent?.read).toBe(true)

            // The LLM provided with a completion result.
            expect(result.inlineCompletionItems.length).toBeGreaterThan(0)
            expect(result.decoratedEditItems.length).toBe(0)

            const texts = result.inlineCompletionItems.map(item => item.insertText)
            expect(texts).toMatchSnapshot()
        }, 10_000)
    })

    describe('autoedit - edits', () => {
        async function getAutoEditSuggestion(
            client: TestClient,
            uri: vscode.Uri,
            position: Position
        ): Promise<AutocompleteEditItem> {
            await client.openFile(uri)

            // Set a small visibility delay for testing purposes.
            const visibilityDelay = 100
            await client.request('testing/autocomplete/setCompletionVisibilityDelay', {
                delay: visibilityDelay,
            })

            // Generate an edit for the current cursor position.
            const result = (await client.request('autocomplete/execute', {
                uri: uri.toString(),
                position,
                triggerKind: 'Automatic',
            })) as AutocompleteResult
            const id = result.decoratedEditItems[0].id

            // Tell completion provider that the completion was shown to the user.
            client.notify('autocomplete/completionSuggested', { completionID: id })

            // Check the initial telemetry event, we haven't yet waited for the visibility timeout
            // so we need to ensure this is only in the "suggested" state
            const initialAutoEditEvent = await client.request('testing/autocomplete/autoeditEvent', {
                completionID: id,
            })
            expect(initialAutoEditEvent?.phase).toBe('suggested')

            // Wait for the completion visibility timeout we use to ensure users read a completion.
            await client.request('testing/autocomplete/awaitPendingVisibilityTimeout', null)

            const autoeditEvent = await client.request('testing/autocomplete/autoeditEvent', {
                completionID: id,
            })
            expect(autoeditEvent?.phase).toBe('read')

            // Expect the result to have at least one edit item
            expect(result.decoratedEditItems.length).toBeGreaterThan(0)
            expect(result.inlineCompletionItems.length).toBe(0)

            // Expect the first item to be an edit
            const editItem = result.decoratedEditItems[0] as AutocompleteEditItem

            return editItem
        }

        describe('client can only render inline diffs', () => {
            const client = TestClient.create({
                workspaceRootUri: workspace.rootUri,
                name: 'autoedit-inline-diffs',
                credentials: TESTING_CREDENTIALS.enterprise,
                extraConfiguration: {
                    experimentalAutoEditEnabled: true,
                },
                capabilities: {
                    ...allClientCapabilitiesEnabled,
                    autoedit: 'enabled',
                    autoeditInlineDiff: 'insertions-and-deletions',
                    autoeditAsideDiff: 'none',
                },
            })

            beforeAll(async () => {
                await client.beforeAll(clientConfiguration)
            })

            afterAll(async () => {
                await client.afterAll()
            })

            it('produces an inline diff for a simple suggestion', async () => {
                const file = workspace.file('src', 'sum-ages.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 4, character: 1 })

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "}

                  export function sumAge(humanA: Person, humanB: Person): number {
                      return humanA.age + humanB.age
                  "
                `)

                const { aside, inline } = result.render

                // No aside options provided (client only supports inline)
                expect(aside.diff).toBeNull()
                expect(aside.image).toBeNull()

                // Inline diff provided
                expect(inline.changes).not.toBeNull()
            }, 10_000)

            it('produces an inline diff for a complex suggestion', async () => {
                const file = workspace.file('src', 'sum-ages-complex-diff.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 6, character: 52 })
                // No completions provided
                // No need to check items array for an edit result

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "export function sumAge(a: Person, b: Person): number {
                      return a.age + b.age
                  }
                  "
                `)

                const { aside, inline } = result.render

                // No aside options provided (client only supports inline)
                expect(aside.diff).toBeNull()
                expect(aside.image).toBeNull()

                // Inline diff provided
                expect(inline.changes).not.toBeNull()
            }, 10_000)
        })

        describe('client can only render aside diffs as images', () => {
            const client = TestClient.create({
                workspaceRootUri: workspace.rootUri,
                name: 'autoedit-aside-images',
                credentials: TESTING_CREDENTIALS.enterprise,
                extraConfiguration: {
                    experimentalAutoEditEnabled: true,
                },
                capabilities: {
                    ...allClientCapabilitiesEnabled,
                    autoedit: 'enabled',
                    autoeditInlineDiff: 'none',
                    autoeditAsideDiff: 'image',
                },
            })

            beforeAll(async () => {
                await client.beforeAll(clientConfiguration)
            })

            afterAll(async () => {
                await client.afterAll()
            })

            it('produces a unified image diff for a simple suggestion', async () => {
                const file = workspace.file('src', 'sum-ages.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 4, character: 1 })

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "}

                  export function sumAge(humanA: Person, humanB: Person): number {
                      return humanA.age + humanB.age
                  "
                `)

                const { aside, inline } = result.render

                // No inline diff provided (client only supports aside)
                expect(inline.changes).toBeNull()

                // No aside diff object provided (client will use the image to show the diff)
                expect(aside.diff).toBeNull()

                // Image is provided
                expect(aside.image).not.toBeNull()
                const { dark, light } = aside.image!
                const darkBuffer = Buffer.from(dark.split(',')[1], 'base64')
                const lightBuffer = Buffer.from(light.split(',')[1], 'base64')
                expect(lightBuffer).toMatchImageSnapshot({
                    customSnapshotIdentifier: 'simple-unified-diff-light',
                })
                expect(darkBuffer).toMatchImageSnapshot({
                    customSnapshotIdentifier: 'simple-unified-diff-dark',
                })
            }, 10_000)

            it('produces a unified image diff for a complex suggestion', async () => {
                const file = workspace.file('src', 'sum-ages-complex-diff.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 6, character: 52 })

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "export function sumAge(a: Person, b: Person): number {
                      return a.age + b.age
                  }
                  "
                `)

                const { aside, inline } = result.render

                // No inline diff provided (client only supports aside)
                expect(inline.changes).toBeNull()

                // No aside diff object provided (client will use the image to show the diff)
                expect(aside.diff).toBeNull()

                // Image is provided
                expect(aside.image).not.toBeNull()
                const { dark, light } = aside.image!
                const darkBuffer = Buffer.from(dark.split(',')[1], 'base64')
                const lightBuffer = Buffer.from(light.split(',')[1], 'base64')
                expect(lightBuffer).toMatchImageSnapshot({
                    customSnapshotIdentifier: 'complex-unified-diff-light',
                })
                expect(darkBuffer).toMatchImageSnapshot({
                    customSnapshotIdentifier: 'complex-unified-diff-dark',
                })
            }, 10_000)
        })

        describe('client can only render aside diffs with their own implementation', () => {
            const client = TestClient.create({
                workspaceRootUri: workspace.rootUri,
                name: 'autoedit-aside-custom',
                credentials: TESTING_CREDENTIALS.enterprise,
                extraConfiguration: {
                    experimentalAutoEditEnabled: true,
                },
                capabilities: {
                    ...allClientCapabilitiesEnabled,
                    autoedit: 'enabled',
                    autoeditInlineDiff: 'none',
                    autoeditAsideDiff: 'diff',
                },
            })

            beforeAll(async () => {
                await client.beforeAll(clientConfiguration)
            })

            afterAll(async () => {
                await client.afterAll()
            })

            it('produces a diff for a simple suggestion', async () => {
                const file = workspace.file('src', 'sum-ages.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 4, character: 1 })

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "}

                  export function sumAge(humanA: Person, humanB: Person): number {
                      return humanA.age + humanB.age
                  "
                `)

                const { aside, inline } = result.render

                // No inline diff provided (client only supports aside)
                expect(inline.changes).toBeNull()

                // No image provided (client will render the aside diff in their own way)
                expect(aside.image).toBeNull()

                // Diff object is provided
                expect(aside.diff).not.toBeNull()
                const { modifiedLines, unchangedLines } = aside.diff!
                // Check the diff has contents, we don't snapshot this as it is quite a large object
                expect(modifiedLines.length).toBeGreaterThan(0)
                expect(unchangedLines.length).toBeGreaterThan(0)
            }, 10_000)

            it('produces a unified image diff for a complex suggestion', async () => {
                const file = workspace.file('src', 'sum-ages-complex-diff.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 6, character: 52 })

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "export function sumAge(a: Person, b: Person): number {
                      return a.age + b.age
                  }
                  "
                `)

                const { aside, inline } = result.render

                // No inline diff provided (client only supports aside)
                expect(inline.changes).toBeNull()

                // No image provided (client will render the aside diff in their own way)
                expect(aside.image).toBeNull()

                // Diff object is provided
                expect(aside.diff).not.toBeNull()
                const { modifiedLines, unchangedLines } = aside.diff!
                // Check the diff has contents, we don't snapshot this as it is quite a large object
                expect(modifiedLines.length).toBeGreaterThan(0)
                expect(unchangedLines.length).toBeGreaterThan(0)
            }, 10_000)
        })

        describe('client can render both inline and aside diffs', () => {
            const client = TestClient.create({
                workspaceRootUri: workspace.rootUri,
                name: 'autoedit-inline-and-aside',
                credentials: TESTING_CREDENTIALS.enterprise,
                extraConfiguration: {
                    experimentalAutoEditEnabled: true,
                },
                capabilities: {
                    ...allClientCapabilitiesEnabled,
                    autoedit: 'enabled',
                    autoeditInlineDiff: 'insertions-and-deletions',
                    autoeditAsideDiff: 'image',
                },
            })

            beforeAll(async () => {
                await client.beforeAll(clientConfiguration)
            })

            afterAll(async () => {
                await client.afterAll()
            })

            it('produces an inline diff for a simple suggestion', async () => {
                const file = workspace.file('src', 'sum-ages.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 4, character: 1 })

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "}

                  export function sumAge(humanA: Person, humanB: Person): number {
                      return humanA.age + humanB.age
                  "
                `)

                const { aside, inline } = result.render

                // No aside options provided (client only supports inline)
                expect(aside.diff).toBeNull()
                expect(aside.image).toBeNull()

                // Inline diff provided
                expect(inline.changes).not.toBeNull()
            }, 10_000)

            it('produces a mix of aside and inline diffs for a complex suggestion', async () => {
                const file = workspace.file('src', 'sum-ages-complex-diff.ts')
                const result = await getAutoEditSuggestion(client, file, { line: 6, character: 52 })

                // Prediction accurately reflects the edit that should be made.
                expect(result.insertText).toMatchInlineSnapshot(`
                  "export function sumAge(a: Person, b: Person): number {
                      return a.age + b.age
                  }
                  "
                `)

                const { aside, inline } = result.render

                // Inline diff provided
                expect(inline.changes).not.toBeNull()

                // Aside diff provided as an image
                expect(aside.image).not.toBeNull()
                const { dark, light } = aside.image!
                const darkBuffer = Buffer.from(dark.split(',')[1], 'base64')
                const lightBuffer = Buffer.from(light.split(',')[1], 'base64')
                expect(lightBuffer).toMatchImageSnapshot({
                    customSnapshotIdentifier: 'complex-mixed-diff-light',
                })
                expect(darkBuffer).toMatchImageSnapshot({
                    customSnapshotIdentifier: 'complex-mixed-diff-dark',
                })
            }, 10_000)
        })
    })
})
