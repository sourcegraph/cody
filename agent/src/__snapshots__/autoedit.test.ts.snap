// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Autoedit > autoedit - completions > autocomplete/execute multiline (non-empty result) 1`] = `
[
  "    const n = arr.length;
    for (let i = 0; i < n - 1; i++) {
        for (let j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                // Swap arr[j] and arr[j + 1]
                const temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
    return arr;",
]
`;
